subprojects/core/src/main/java/org/gradle/api/provider/PropertyState.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.provider;
+
+import org.gradle.api.Incubating;
+
+/**
+ * A {@code Provider} representation for capturing property state. The value can be provided by using the method {@code set()}.
+ * <p>
+ * You can create a {@code PropertyState} instance using the method {@link org.gradle.api.Project#property(java.lang.Class)}.
+ *
+ * @param <T> Type of value represented by property state
+ * @since 3.5
+ */
+@Incubating
+public interface PropertyState<T> extends Provider<T> {
```

Comment:
Would you mind expanding on the origin of this class name a bit? I find it a bit unclear and could use an explanation.